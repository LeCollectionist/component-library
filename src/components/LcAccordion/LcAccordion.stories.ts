import LcButton from '../LcButton/index'
import LcAccordion from './index'

export default {
  title: 'Example/LcAccordion',
  component: LcAccordion,
}

const Template = (args: any) => ({
  components: { LcAccordion, LcButton },
  setup() {
    return { args }
  },
  data() {
    return { open: args.modelValue }
  },
  template: `
    <lc-accordion v-bind="args" v-model="open">
      <template #body>
        My content
      </template>
    </lc-accordion>
  `,
})

const TemplateSlotActionsAfter = (args: any) => ({
  components: { LcAccordion, LcButton },
  setup() {
    return { args }
  },
  data() {
    return { open: args.modelValue }
  },
  template: `
    <lc-accordion v-bind="args" v-model="open">
      <template #content-after>
        <lc-button>button after</lc-button>
      </template>
      <template #body>
        My content
      </template>
    </lc-accordion>
  `,
})

const TemplateSlotActionsBefore = (args: any) => ({
  components: { LcAccordion, LcButton },
  setup() {
    return { args }
  },
  data() {
    return { open: args.modelValue }
  },
  template: `
    <lc-accordion v-bind="args" v-model="open">
      <template #content-before>
        <lc-button>button before</lc-button>
      </template>
      <template #body>
        My content
      </template>
    </lc-accordion>
  `,
})

export const Base = Template.bind({}) as any
Base.args = {
  title: '1. House information',
  modelValue: true,
}

export const HaveToogle = Template.bind({}) as any
HaveToogle.args = {
  ...Base.args,
  haveToggle: true,
}

export const HaveToogleButClose = Template.bind({}) as any
HaveToogleButClose.args = {
  ...HaveToogle.args,
  modelValue: false,
}

export const HaveButtonRight = TemplateSlotActionsAfter.bind({}) as any
HaveButtonRight.args = {
  title: '1. House information',
  modelValue: true,
}

export const HaveButtonRightAndToggle = TemplateSlotActionsAfter.bind({}) as any
HaveButtonRightAndToggle.args = {
  ...HaveButtonRight.args,
  haveToggle: true,
}

export const HaveButtonLeft = TemplateSlotActionsBefore.bind({}) as any
HaveButtonLeft.args = {
  title: '1. House information',
  modelValue: true,
}

export const HaveButtonLeftAndToggle = TemplateSlotActionsBefore.bind({}) as any
HaveButtonLeftAndToggle.args = {
  ...HaveButtonLeft.args,
  haveToggle: true,
}

export const HaveNoBodyPadding = Template.bind({}) as any
HaveNoBodyPadding.args = {
  ...Base.args,
  hasBodyPadding: false,
}
