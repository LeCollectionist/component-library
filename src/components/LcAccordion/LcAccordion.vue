<script lang="ts">
import { defineComponent } from 'vue'

import LcButton from '../LcButton'
import LcIcon from '../LcIcon'

type Height = number | 'auto'

export default defineComponent({
  components: {
    LcButton,
    LcIcon,
  },
  props: {
    duration: {
      type: Number,
      default: 250,
    },
    hasBodyPadding: {
      type: Boolean,
      default: true,
    },
    haveToggle: {
      type: Boolean,
      default: false,
    },
    modelValue: {
      type: Boolean,
      default: true,
    },
    title: {
      type: String,
      default: '',
    },
  },
  emits: ['update:model-value'],
  computed: {
    containerStyle(): Record<string, string> {
      return {
        transitionDuration: `${this.duration}ms`,
      }
    },
  },
  mounted() {
    if (this.modelValue) {
      const el = this.$refs['lc-accordion-body'] as HTMLElement
      this.setContainerHeight('auto', el)
    }
  },
  methods: {
    getContentHeight() {
      const wrapper = this.$refs.wrapper as HTMLElement
      return wrapper?.getBoundingClientRect().height || 0
    },
    onEnter(el: HTMLElement) {
      this.setContainerHeight(this.getContentHeight(), el)
      el.style.overflow = 'hidden'
    },
    onAfterEnter(el: HTMLElement) {
      this.setContainerHeight('auto', el)
      el.style.overflow = 'visible'
    },
    onBeforeLeave(el: HTMLElement) {
      this.setContainerHeight(this.getContentHeight(), el)
      el.style.overflow = 'hidden'
    },
    onLeave(el: HTMLElement) {
      this.setContainerHeight(0, el)
    },
    setContainerHeight(height: Height, el: HTMLElement) {
      el.style.height = typeof height === 'number' ? `${height}px` : height
    },
    toggle() {
      this.$emit('update:model-value', !this.modelValue)
    },
  },
})
</script>

<template>
  <div class="lc-accordion">
    <div
      class="lc-accordion-header"
      data-testid="lc-accordion-header"
    >
      <div>
        <span
          v-if="title"
          class="lc-accordion-header-title"
          data-testid="lc-accordion-header-title"
        >
          {{ title }}
        </span>
        <slot name="content-before" />
      </div>
      <div
        v-if="$slots['content-after'] || haveToggle"
        class="lc-accordion-header-right"
      >
        <slot name="content-after" />
        <LcButton
          v-if="haveToggle"
          class="lc-accordion-header-toggle-button"
          data-testid="lc-accordion-header-toggle-button"
          variant="icon"
          size="xs"
          @click="toggle"
        >
          <slot name="button-toggle-icon">
            <LcIcon
              class="lc-accordion-header-toggle-button-icon"
              :name="modelValue ? 'top-chevron' : 'bottom-chevron'"
              size="xxs"
            />
          </slot>
        </LcButton>
      </div>
    </div>

    <transition
      @enter="onEnter"
      @after-enter="onAfterEnter"
      @before-leave="onBeforeLeave"
      @leave="onLeave"
    >
      <div
        v-show="modelValue"
        ref="lc-accordion-body"
        class="lc-accordion-body"
        data-testid="lc-accordion-body"
        :style="containerStyle"
      >
        <div ref="wrapper" class="lc-accordion-body-wrapper" :class="[{ 'lc-accordion-body-wrapper--withpadding': hasBodyPadding }]">
          <slot name="body" />
        </div>
      </div>
    </transition>
  </div>
</template>

<style>
.lc-accordion {
  @apply bg-white shadow-md;
}

.lc-accordion-header {
  @apply px-9 py-4 border-b border-gray-300 flex items-center justify-between;
}
.lc-accordion-header-title {
  @apply text-xl font-medium mr-4;
}
.lc-accordion-header-toggle-button {
  @apply ml-4;
}
.lc-accordion-header-toggle-button-icon {
  @apply text-white;
}
.lc-accordion-header-right {
  @apply flex items-center;
}

.lc-accordion-body {
   @apply h-0 ease-in-out transition-height;
}
.lc-accordion-body-wrapper {
  @apply w-full;
}
.lc-accordion-body-wrapper--withpadding {
  @apply px-9 py-5;
}
</style>
