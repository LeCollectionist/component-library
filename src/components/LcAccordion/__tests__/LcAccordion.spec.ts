import { mount } from '@vue/test-utils'
import LcAccordion from '../LcAccordion'

let wrapper: any

afterEach(() => {
  wrapper?.unmount()
})

describe('LcAccordion', () => {
  it('is a Vue instance', () => {
    wrapper = mount(LcAccordion)
    expect(wrapper.vm).toBeTruthy()
  })

  describe('Default behavior', () => {
    beforeEach(() => {
      wrapper = mount(LcAccordion)
    })

    it('should render title', async () => {
      await wrapper.setProps({ title: 'My title' })

      const title = wrapper.find('[data-testid="lc-accordion-header-title"]')

      expect(title.exists()).toBeTruthy()
      expect(title.text()).toBe('My title')
    })

    it('shouldn\'t render title', () => {
      const title = wrapper.find('[data-testid="lc-accordion-header-title"]')

      expect(title.exists()).toBeFalsy()
    })

    it('should body is hide', async () => {
      await wrapper.setProps({ modelValue: false })

      const body = wrapper.find('[data-testid="lc-accordion-body"]')
      expect(body.isVisible()).toBe(false)
    })

    it('should body is visible', () => {
      const body = wrapper.find('[data-testid="lc-accordion-body"]')
      expect(body.isVisible()).toBe(true)
    })

    it('should set bodypadding class on body-wrapper', () => {
      const bodyWrapper = wrapper.find('[data-testid="lc-accordion-body"] > div')
      expect(bodyWrapper.classes()).toContain('lc-accordion-body-wrapper--withpadding')
    })

    it('should not set bodypadding class on body-wrapper', async () => {
      await wrapper.setProps({ hasBodyPadding: false })

      const bodyWrapper = wrapper.find('[data-testid="lc-accordion-body"] > div')
      expect(bodyWrapper.classes()).not.toContain('lc-accordion-body-wrapper--withpadding')
    })
  })

  describe('Toggle button', () => {
    it('shouldn\'t render button toggle', () => {
      wrapper = mount(LcAccordion)

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')

      expect(button.exists()).toBeFalsy()
    })

    it('should render button toggle', () => {
      wrapper = mount(LcAccordion, {
        props: { haveToggle: true },
      })

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')

      expect(button.exists()).toBeTruthy()
    })

    it('should emit update:modelValue with value false', async () => {
      wrapper = mount(LcAccordion, {
        props: { haveToggle: true },
      })

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')
      await button.trigger('click')

      expect(wrapper.emitted('update:modelValue')).toBeTruthy()
      expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual([false])
    })

    it('should emit update:modelValue with value true', async () => {
      wrapper = mount(LcAccordion, {
        props: { haveToggle: true, modelValue: false },
      })

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')
      await button.trigger('click')

      expect(wrapper.emitted('update:modelValue')).toBeTruthy()
      expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual([true])
    })

    it('should render body accordion visible then hide', async () => {
      const wrapper = mount({
        data() {
          return { open: true }
        },
        template: '<lc-accordion v-model="open" :have-toggle="true" />',
        components: { LcAccordion },
      })

      const lcAccordion = wrapper.findComponent(LcAccordion)
      const bodyBeforeClick = lcAccordion.find('[data-testid="lc-accordion-body"]')

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')
      await button.trigger('click')
      const bodyAfterClick = lcAccordion.find('[data-testid="lc-accordion-body"]')

      expect(bodyBeforeClick.isVisible()).toBe(true)
      expect(bodyAfterClick.isVisible()).toBe(false)
    })

    it('should render body accordion hide then visible', async () => {
      const wrapper = mount({
        data() {
          return { open: false }
        },
        template: '<lc-accordion v-model="open" :have-toggle="true" />',
        components: { LcAccordion },
      })

      const lcAccordion = wrapper.findComponent(LcAccordion)
      const bodyBeforeClick = lcAccordion.find('[data-testid="lc-accordion-body"]')

      const button = wrapper.find('[data-testid="lc-accordion-header-toggle-button"]')
      await button.trigger('click')
      const bodyAfterClick = lcAccordion.find('[data-testid="lc-accordion-body"]')

      expect(bodyBeforeClick.isVisible()).toBe(false)
      expect(bodyAfterClick.isVisible()).toBe(true)
    })
  })

  describe('Before slot', () => {
    beforeEach(() => {
      wrapper = mount(LcAccordion, {
        slots: { 'content-before': '<button data-testid="button-before-slot">test</button>' },
      })
    })

    it('should render content before slot', () => {
      const contentSlot = wrapper.find('[data-testid="button-before-slot"]')
      expect(contentSlot.exists()).toBeTruthy()
    })

    it('should render HTML header with before slot', () => {
      const header = wrapper.find('[data-testid="lc-accordion-header"]')

      expect(header.html()).toMatchSnapshot()
    })

    it('should render HTML header with button toogle + befor slot', async () => {
      await wrapper.setProps({ haveToggle: true })

      const header = wrapper.find('[data-testid="lc-accordion-header"]')
      expect(header.html()).toMatchSnapshot()
    })
  })

  describe('After slot', () => {
    beforeEach(() => {
      wrapper = mount(LcAccordion, {
        slots: { 'content-after': '<button data-testid="button-after-slot">test</button>' },
      })
    })

    it('should render content after slot', () => {
      const contentSlot = wrapper.find('[data-testid="button-after-slot"]')
      expect(contentSlot.exists()).toBeTruthy()
    })

    it('should render HTML header with after slot', () => {
      const header = wrapper.find('[data-testid="lc-accordion-header"]')

      expect(header.html()).toMatchSnapshot()
    })

    it('should render HTML header with button toogle + after slot', async () => {
      await wrapper.setProps({ haveToggle: true })

      const header = wrapper.find('[data-testid="lc-accordion-header"]')
      expect(header.html()).toMatchSnapshot()
    })
  })

  describe('Without slots', () => {
    beforeEach(() => {
      wrapper = mount(LcAccordion, {
        props: { title: 'My title' },
      })
    })

    it('should render HTML header (only title)', async () => {
      const header = wrapper.find('[data-testid="lc-accordion-header"]')

      expect(header.html()).toMatchSnapshot()
    })

    it('should render HTML header with button toggle', async () => {
      await wrapper.setProps({ haveToggle: true })

      const header = wrapper.find('[data-testid="lc-accordion-header"]')

      expect(header.html()).toMatchSnapshot()
    })
  })
})
