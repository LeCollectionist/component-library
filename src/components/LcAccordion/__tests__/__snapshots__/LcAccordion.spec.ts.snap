// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LcAccordion After slot should render HTML header with after slot 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><!--v-if--></div><div class=\\"lc-accordion-header-right\\"><button data-testid=\\"button-after-slot\\">test</button><!--v-if--></div></div>"`;

exports[`LcAccordion After slot should render HTML header with button toogle + after slot 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><!--v-if--></div><div class=\\"lc-accordion-header-right\\"><button data-testid=\\"button-after-slot\\">test</button><button class=\\"lc-icon--primary lc-icon--xs lc-icon lc-icon--font-bold lc-accordion-header-toggle-button\\" data-testid=\\"lc-accordion-header-toggle-button\\" type=\\"button\\"><!--v-if--><span class=\\"icon icon-top-chevron icon-xxs lc-accordion-header-toggle-button-icon\\"></span></button></div></div>"`;

exports[`LcAccordion Before slot should render HTML header with before slot 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><!--v-if--><button data-testid=\\"button-before-slot\\">test</button></div><!--v-if--></div>"`;

exports[`LcAccordion Before slot should render HTML header with button toogle + befor slot 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><!--v-if--><button data-testid=\\"button-before-slot\\">test</button></div><div class=\\"lc-accordion-header-right\\"><button class=\\"lc-icon--primary lc-icon--xs lc-icon lc-icon--font-bold lc-accordion-header-toggle-button\\" data-testid=\\"lc-accordion-header-toggle-button\\" type=\\"button\\"><!--v-if--><span class=\\"icon icon-top-chevron icon-xxs lc-accordion-header-toggle-button-icon\\"></span></button></div></div>"`;

exports[`LcAccordion Without slots should render HTML header (only title) 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><span class=\\"lc-accordion-header-title\\" data-testid=\\"lc-accordion-header-title\\">My title</span></div><!--v-if--></div>"`;

exports[`LcAccordion Without slots should render HTML header with button toggle 1`] = `"<div class=\\"lc-accordion-header\\" data-testid=\\"lc-accordion-header\\"><div><span class=\\"lc-accordion-header-title\\" data-testid=\\"lc-accordion-header-title\\">My title</span></div><div class=\\"lc-accordion-header-right\\"><button class=\\"lc-icon--primary lc-icon--xs lc-icon lc-icon--font-bold lc-accordion-header-toggle-button\\" data-testid=\\"lc-accordion-header-toggle-button\\" type=\\"button\\"><!--v-if--><span class=\\"icon icon-top-chevron icon-xxs lc-accordion-header-toggle-button-icon\\"></span></button></div></div>"`;
